cmd/clusterctl/pkg/log/logger.go

```diff
@@ -0,0 +1,307 @@
+
+// NewOptions carries the options supported by NewLogger
+type NewOptions struct {
+	ch        LogChannel
+	threshold int
+}
+
+// Option is a configuration option supplied to NewLogger
+type Option func(*NewOptions)
+
+// WithThreshold implements a New Option that allows to set the threshold level for a new logger.
+// The logger will write only log messages with a level/V(x) equal or higher to the threshold.
+func WithThreshold(threshold int) Option {
+	return func(c *NewOptions) {
+		c.threshold = threshold
+	}
+}
+
+// WithChannel implements a New Option that allows to set the channel where LogEntry can be sent for a new logger.
+// Please note that the logger is designed to not drop messages, so the channel should be ready to receive messages otherwise
+// writing on the logger will block.
+func WithChannel(ch LogChannel) Option {
+	return func(c *NewOptions) {
+		c.ch = ch
+	}
+}
+
+// NewLogger returns a new instance of the clusterctl.
+func NewLogger(options ...Option) logr.Logger {
+	cfg := &NewOptions{}
+	for _, o := range options {
+		o(cfg)
+	}
+
+	return &Logger{
+		threshold: cfg.threshold,
+		ch:        cfg.ch,
+		level:     0,
+		prefix:    "",
+		values:    nil,
+	}
+}
+
+// Logger defines a clusterctl friendly logr.Logger
+type Logger struct {
+	threshold int
+	ch        LogChannel
+	level     int
+	prefix    string
+	values    []interface{}
+}
+
+var _ logr.Logger = &Logger{}
+
+// Enabled tests whether this Logger is enabled.
+func (l *Logger) Enabled() bool {
+	return l.level >= l.threshold
+}
+
+// Info logs a non-error message with the given key/value pairs as context.
+func (l *Logger) Info(msg string, kvs ...interface{}) {
+	if l.Enabled() {
+		values := copySlice(l.values)
+		values = append(values, kvs...)
+		values = append(values, "msg", msg)
+		l.write(values)
+	}
+}
+
+// Error logs an error message with the given key/value pairs as context.
+func (l *Logger) Error(err error, msg string, kvs ...interface{}) {
+	values := copySlice(l.values)
+	values = append(values, kvs...)
+	values = append(values, "msg", msg, "error", err)
+	l.write(values)
+}
+
+// V returns an InfoLogger value for a specific verbosity level.
+func (l *Logger) V(level int) logr.InfoLogger {
+	nl := l.clone()
+	nl.level = level
+	return nl
+}
+
+// WithName adds a new element to the logger's name.
+func (l *Logger) WithName(name string) logr.Logger {
+	nl := l.clone()
+	if len(l.prefix) > 0 {
+		nl.prefix = l.prefix + "/"
+	}
+	nl.prefix += name
+	return nl
+}
+
+// WithValues adds some key-value pairs of context to a logger.
+func (l *Logger) WithValues(kvList ...interface{}) logr.Logger {
+	nl := l.clone()
+	nl.values = append(nl.values, kvList...)
+	return nl
+}
+
+func (l *Logger) write(values []interface{}) {
+	entry := LogEntry{
+		Prefix: l.prefix,
+		Level:  l.level,
+		Values: values,
+	}
+	if l.ch != nil {
+		l.ch <- entry
+		return
+	}
+	f, err := Flatten(entry)
+	if err != nil {
+		panic(err)
+	}
+	fmt.Println(f)
```

Comment:
Doesn't have to be in this PR. But we could pass in an io.Writer to the logger and use that. 
This would also help with testing because I noticed that we aren't really testing public functions of the package such as `logger.Error` except for `Flatten` which happens to be internal to `write`.